packages/medusa-payment-paypal/src/services/paypal-provider.ts

```diff
@@ -0,0 +1,386 @@
+      }
+      const isAlreadyCaptured = purchase_units.some(
+        (pu) => pu.payments.captures?.length
+      )
+
+      if (isAlreadyCaptured) {
+        const payments = purchase_units[0].payments
+
+        const payId = payments.captures[0].id
+        const request = new PayPal.payments.CapturesRefundRequest(payId)
+        await this.paypal_.execute(request)
+      } else {
+        const id = purchase_units[0].payments.authorizations[0].id
+        const request = new PayPal.payments.AuthorizationsVoidRequest(id)
+        await this.paypal_.execute(request)
+      }
+
+      return (await this.retrievePayment(
+        paymentSessionData
+      )) as unknown as PaymentProcessorSessionResponse["session_data"]
+    } catch (error) {
+      return this.buildError("An error occurred in cancelPayment", error)
+    }
+  }
+
+  async capturePayment(
+    context: PaymentProcessorContext
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    const paymentSessionData = context.paymentSessionData
+    const { purchase_units } = paymentSessionData as {
+      purchase_units: PurchaseUnits
+    }
+
+    const id = purchase_units[0].payments.authorizations[0].id
+
+    try {
+      const request = new PayPal.payments.AuthorizationsCaptureRequest(id)
+      await this.paypal_.execute(request)
+      return this.retrievePayment(paymentSessionData)
+    } catch (error) {
+      return this.buildError("An error occurred in capturePayment", error)
+    }
+  }
+
+  /**
+   * Paypal does not provide such feature
+   * @param paymentSessionData
+   */
+  async deletePayment(
+    paymentSessionData: Record<string, unknown>
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    return paymentSessionData
+  }
+
+  async refundPayment(
+    paymentSessionData: Record<string, unknown>,
+    refundAmount: number
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    const { purchase_units } = paymentSessionData as {
+      purchase_units: PurchaseUnits
+    }
+
+    try {
+      const purchaseUnit = purchase_units[0]
+      const payments = purchaseUnit.payments
+      const isAlreadyCaptured = purchase_units.some(
+        (pu) => pu.payments.captures?.length
+      )
+
+      if (!isAlreadyCaptured) {
+        throw new Error("The order has not yet been captured. Unable to refund")
+      }
+
+      const paymentId = payments.captures[0].id
+      const currencyCode = purchaseUnit.amount.currency_code
+      const request = new PayPal.payments.CapturesRefundRequest(paymentId)
+
+      request.requestBody({
+        amount: {
+          currency_code: currencyCode,
+          value: roundToTwo(
+            humanizeAmount(refundAmount, currencyCode),
+            currencyCode
+          ),
+        },
+      })
+
+      await this.paypal_.execute(request)
+
+      return this.retrievePayment(paymentSessionData)
```

Comment:
todo: Add missing await
```suggestion
      return (await this.retrievePayment(
        paymentSessionData
      )) as unknown as PaymentProcessorSessionResponse["session_data"]
```